validator/compute_zone.rego

```diff
@@ -0,0 +1,64 @@
+#
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPComputeZoneV1
+
+import data.validator.gcp.lib as lib
+
+#####################################
+# Find Compute Asset Zone Violations
+#####################################
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	lib.get_constraint_params(constraint, params)
+
+	# Verify that resource is Instance or Disk
+	asset := input.asset
+	{asset.asset_type} == {asset.asset_type} & {"compute.googleapis.com/Instance", "compute.googleapis.com/Disk"}
+
+	# Check if resource is in exempt list
+	exempt_list := params.exemptions
+	matches := {asset.name} & cast_set(exempt_list)
+	count(matches) = 0
+
+	# Check that zone is in allowlist/denylist
+	target_zones := params.zones
+	asset_zone_tokens := split(asset.resource.data.zone, "/")
+	asset_zone := asset_zone_tokens[8]
```

Comment:
optional: maybe `asset_zone_tokens[count(asset_zone_tokens) - 1]` would be more resilient?

Reply:
I had originally tried that, but it wasn't working for me for some reason, and I just chalked it up to a poor understanding of Rego. It must've been a side effect of something else I was doing at the time, because it's working fine now. Done.